fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -85,122 +43,658 @@ def dloss(input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
+        must have been added previously to the graph and are listed with argument inputs.
+        """
+        self._set_inputs(module, [(self.modules_list.index(m), 0) for m in inputs])
+
+    def fan_out(self, module: PipelineModule, outputs: List[PipelineModule]) -> None:
+        """Feeds outputs of a previously added module to modules specified by argument 'outputs' (so
+        'module' should have at least 'len(outputs)' outputs.
+        Modules in the list 'outputs' are added to the graph if they have not been added previously.
+        """
+        mi = self.modules_list.index(module)
+        for i, m in enumerate(outputs):
+            self._set_inputs(m, [(mi, i)])
+
+    @dataclass
+    class OutputConsumerIndex:
+        """A data class for representating a consumer of an output of a module."""
+
+        consumer_idx: int  # index of the consumer module
+        consumer_input_idx: int  # indicating which input of the consumer module
+        producer_output_idx: int  # indicating which output of the producer module
+
+    @staticmethod
+    def _is_model_input(input_desc: Tuple[int, int]) -> bool:
+        """Checks if an entry in self.inputs refers to a model input"""
+        return input_desc[0] < 0
+
+    def _compile(self) -> None:
+        """Precomputes self.model_input_consumers and self.output_consumers for internal use by the pipleine
+        class. These two lists show consumers of inputs to the model, and outputs of each module of
+        the graph. Each consumer is a pair (i, j) which stands for the j'th input to the i'th module
+        in the graph.
+        """
+
+        # TODO: We need to make sure following conditions hold before preparing the graph for the pipeline:
+        #   * the graph has a least one module, and is connected.
+        #   * num_inputs and num_outputs for modules matche list of connections defined in the graph.
+        #   * all inputs to a module should come from model input, or modules with smaller index in
+        #     the graph. This condition is used in implementaion of DistributedPipeline.forward. Even
+        #     if we relax this condition, still need to make sure the graph is acyclic.
+
+        m = len(self.modules_list)
+        self.output_consumers: List[List[PipelineModulesGraph.OutputConsumerIndex]] = [[] for _ in range(m)]
+        self.model_input_consumers = []
+        for i, input in enumerate(self.inputs):
+            assert input is not None
+            for j, input_item in enumerate(input):
+                if not self._is_model_input(input_item):
+                    self.output_consumers[input_item[0]].append(
+                        PipelineModulesGraph.OutputConsumerIndex(i, j, input_item[1])
+                    )
+                else:
+                    self.model_input_consumers.append((i, j, input_item[1]))
+
+    def _trace_modules(self, module_idx: int) -> List[int]:
```

Comment:
It feels like the code would be much simpler if we just used the modules directly instead of adding the indirection of index tables.

Reply:
It is not so easy to get rid of indexes. For example self.output_consumers is indexed by module index, and we store indexes of modules in self.inputs and in output_consumers. We change all that to module object ids, but let's do that as a separate PR, since it is not going to be a small change (if we think that kind of change helps).

Reply:
My primary concern is readability. I still find it very challenging to read through the code. For me, the i and j variables and arrays of objects are what make the code most challenging. If they could be hidden behind a lower-level API or removed entirely, it would make the code much easier to read and understand. Each iteration of review is still very challenging. I would like to see the code get a stage where it flows naturally when reading.

Reply:
I see, so I merge other changes in #591 to this branch, and do a major revision to reduce from complexity

Reply:
I went over all data structures used in the PR changed them to help readability. Also splitted the code to smaller modules. Please have another look.
I suggest you first have a look at #591 (which is based on changes that you have already reviewed), because in this PR the files are now splitted and it's hard to understand how they have changed.